core/app/models/spree/address.rb

```diff
@@ -18,19 +18,60 @@ class Address < Spree::Base
     alias_attribute :first_name, :firstname
     alias_attribute :last_name, :lastname
 
+    DB_ONLY_ATTRS = %w(id updated_at created_at)
+
+    scope :with_values, ->(attributes) do
+      where(value_attributes(attributes))
+    end
+
     def self.build_default
-      country = Spree::Country.find(Spree::Config[:default_country_id]) rescue Spree::Country.first
-      new(country: country)
+      new(country: Spree::Country.default)
     end
 
     def self.default(user = nil, kind = "bill")
-      if user && user_address = user.send(:"#{kind}_address")
-        user_address.dup
+      ActiveSupport::Deprecation.warn("Address.default is deprecated. Use User.default_address or Address.build_default", caller)
+      if user
+        user.send(:"#{kind}_address") || build_default
       else
         build_default
       end
     end
 
+    # @return [Address] an equal address already in the database or a newly created one
+    def self.factory(attributes)
+      full_attributes = value_attributes(column_defaults, attributes)
+      find_or_create_by(full_attributes)
+    end
+
+    # @return [Address] address from existing address plus new_attributes as diff
+    # @note, this may return existing_address if there are no changes to value equality
+    def self.immutable_merge(existing_address, new_attributes)
+      return self.new(value_attributes(new_attributes)) unless existing_address
```

Comment:
Just noticed this -- if existing_address is nil don't we still want to search the DB for an address with `new_attributes`?  Otherwise we will end up creating duplicates I think.  e.g., given an order with order.ship_address==nil, doing this: `order.ship_address_attributes = {...}` (which uses `immutable_address`) will always create a new address, even if a matching one already exists.